sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -287,3 +290,231 @@ case class ArrayContains(left: Expression, right: Expression)
+      } else {
+        val elements = inputs.flatMap(_.asInstanceOf[ArrayData].toObjectArray(elementType))
+        new GenericArrayData(elements)
+      }
+  }
+
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val evals = children.map(_.genCode(ctx))
+    val args = ctx.freshName("args")
+
+    val inputs = evals.zipWithIndex.map { case (eval, index) =>
+      s"""
+        ${eval.code}
+        if (!${eval.isNull}) {
+          $args[$index] = ${eval.value};
+        }
+      """
+    }
+
+    val (concatenator, initCode) = dataType match {
+      case BinaryType =>
+        (classOf[ByteArray].getName, s"byte[][] $args = new byte[${evals.length}][];")
+      case StringType =>
+        ("UTF8String", s"UTF8String[] $args = new UTF8String[${evals.length}];")
+      case ArrayType(elementType, _) =>
+        val arrayConcatClass = if (CodeGenerator.isPrimitiveType(elementType)) {
+          genCodeForPrimitiveArrayConcat(ctx, elementType)
+        } else {
+          genCodeForComplexArrayConcat(ctx, elementType)
+        }
+        (arrayConcatClass, s"ArrayData[] $args = new ArrayData[${evals.length}];")
+    }
+    val codes = ctx.splitExpressionsWithCurrentInputs(
+      expressions = inputs,
+      funcName = "valueConcat",
+      extraArguments = (s"${javaType}[]", args) :: Nil)
+    ev.copy(s"""
+      $initCode
+      $codes
+      ${javaType} ${ev.value} = $concatenator.concat($args);
+      boolean ${ev.isNull} = ${ev.value} == null;
+    """)
+  }
+
+  private def genCodeForNumberOfElements(ctx: CodegenContext) : (String, String) = {
+    val tempVariableName = ctx.freshName("tempNumElements")
+    val numElementsConstant = ctx.freshName("numElements")
+    val assignments = (0 until children.length)
+      .map(idx => s"$tempVariableName[0] += args[$idx].numElements();")
+
+    val assignmentSection = ctx.splitExpressions(
+      expressions = assignments,
+      funcName = "complexArrayConcat",
+      arguments = Seq((s"${javaType}[]", "args"), ("int[]", tempVariableName)))
+
+    (s"""
+        |int[] $tempVariableName = new int[]{0};
+        |$assignmentSection
+        |final int $numElementsConstant = $tempVariableName[0];
+      """.stripMargin,
+     numElementsConstant)
+  }
+
+  private def nullArgumentProtection(ctx: CodegenContext) : String = {
+    val isNullVariable = ctx.freshName("isArrayNull")
+    val assignments = children
+      .zipWithIndex
+      .filter(_._1.nullable)
+      .map(ci => s"$isNullVariable[0] |= args[${ci._2}] == null;")
+
+    if (assignments.length > 0) {
+      val assignmentSection = ctx.splitExpressions(
+        expressions = assignments,
+        funcName = "isNullArrayConcat",
+        arguments = Seq((s"${javaType}[]", "args"), ("boolean[]", isNullVariable)))
+
+      s"""
+         |boolean[] $isNullVariable = new boolean[]{false};
+         |$assignmentSection;
+         |if ($isNullVariable[0]) return null;
+       """.stripMargin
```

Comment:
I guess we can simply use for-loop here?

```java
for (int $idx = 0; $idx < ${children.length}; $idx++) {
  if (args[$idx] == null) {
    return null;
  }
}
```

We can return as soon as we found `null` in this case.